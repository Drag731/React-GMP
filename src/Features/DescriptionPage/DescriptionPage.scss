.description-page {
    max-width: 1000px;
    margin:0 auto;
    &__header {
      padding: 10px 30px;
      background-color: pink;
    }
  &__genre-result {
    padding: 10px 30px;
    background-color: lightgrey;
  }
}

.b-header__logo-wrap {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.b-header-info {
  display: flex;
  &__img {
    max-width: 300px;
    margin-right: 30px;
    &-pic {
      width: 100%;
    }
  }
  &__about {
    padding: 10px 15px;
    &-title {
      font-size: 30px;
      margin-bottom: 10px;
      color: red;
    }
    &-year {
      color: white;
      padding:  15px 0;
    }
  }
}

.b-year__release-date {
  margin-right: 30px;
}