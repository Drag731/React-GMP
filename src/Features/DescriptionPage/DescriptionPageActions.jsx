export const ADD_MOVIE = 'ADD_MOVIE';
export const CHANGE_LOADING = 'CHANGE_LOADING';
export const GO_TO_SEARCH = 'GO_TO_SEARCH';

export const receiveMovieDB = (id) => {
    return (dispatch, getState, api) => {
        api(`movies/${id}`)
            .then(response => {
                dispatch(receiveMovieState(response.data));
            })
            .catch(() => {
                    throw new Error('I crashed!');
            })
    };
};

export const receiveMovieState = payload => ({
    type: ADD_MOVIE,
    payload,
});

export const changeIsLoading = () => ({
    type: CHANGE_LOADING,
});

export const goToSearch = () => ({
    type: GO_TO_SEARCH,
});


