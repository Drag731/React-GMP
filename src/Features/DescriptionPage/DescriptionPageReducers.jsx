import {
    ADD_MOVIE,
    CHANGE_LOADING
} from './DescriptionPageActions';

import {
    ADD_MOVIES,
    GO_TO_DESCRIPTION
} from '../MainPage/MainPageActions';

const initialState = {
    genres: [],
    isLoadingMovie: true,
    movie: {},
    movies: []
};

const DescriptionPageReducer = (state = initialState, action) => {
    switch (action.type) {

        case ADD_MOVIE: {
            return {
                ...state,
                isLoadingMovie: false,
                movie: action.payload,
                genres: action.payload.genres
            };
        }

        case ADD_MOVIES: {
            return {
                ...state,
                movies: action.payload.data,
            };
        }

        case CHANGE_LOADING: {
            return {
                ...state,
                isLoadingMovie: true
            };
        }

        case GO_TO_DESCRIPTION: {
            return {
                ...state,
                isLoadingMovie: true,
            };
        }

        default: {
            return state;
        }
    }
};

export const getIsLoadingMovie = state => state.description.isLoadingMovie;
export const getMovie = state => state.description.movie;
export const getMovies = state => state.description.movies;
export const getGenre = state => state.description.genres;

export default DescriptionPageReducer;