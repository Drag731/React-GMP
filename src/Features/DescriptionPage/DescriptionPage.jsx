import React, { Component } from 'react';
import { connect } from 'react-redux';

import ErrorBoundary from '../../components/ErrorBoundary';
import HeaderLogo from '../../components/HeaderLogo/HeaderLogo';
import MovieItemsBlock from '../../components/MovieItemsBlock/MovieItemsBlock';
import HeaderInfo from './components/HeaderInfo/HeaderInfo';

import { receiveMovieDB, changeIsLoading, goToSearch } from './DescriptionPageActions';

import { getMovie, getMovies, getGenre, getIsLoadingMovie } from './DescriptionPageReducers';

import { receiveMoviesDB } from '../MainPage/MainPageActions';

import './DescriptionPage.scss'

const mapStateToProps = state => ({
    movie: getMovie(state),
    genres: getGenre(state),
    isLoadingMovie: getIsLoadingMovie(state),
    movies: getMovies(state)
});

const mapDispatchToProps = {
    receiveMovieDB: (id) => receiveMovieDB(id),
    receiveMoviesDB: (q) => receiveMoviesDB(q),
    changeIsLoading: () => changeIsLoading(),
    goToSearch: () => goToSearch()
};

class DescriptionPage extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            isLoadingMovie: true
        };

        this.handlePop = this.handlePop.bind(this);
    }

    componentDidMount() {
        this.props.receiveMovieDB(this.props.match.params.id);
        this.props.receiveMoviesDB()
    }

    componentWillMount() {
        addEventListener("popstate", this.handlePop)
    }

    componentWillUnmount() {
        removeEventListener("popstate", this.handlePop)
    }

    handlePop() {
        console.log("haddle.pop", this.props);
        this.props.changeIsLoading();
        this.props.receiveMovieDB(this.props.match.params.id);
        this.props.receiveMoviesDB(`?${this.props.genres[0]}`)
    };

    componentWillReceiveProps(nextProps) {
        const {id} = nextProps.match.params;
        if(id && this.props.match.params.id !== id) {
            this.props.changeIsLoading();
            this.props.receiveMovieDB(id);
            this.props.receiveMoviesDB(`?${nextProps.genres[0]}`)
        }
    };

    handleGoSearchPage = () => { this.props.goToSearch()};

    render() {
        return (
            <ErrorBoundary>
                <React.Fragment>
                    {this.props.isLoadingMovie ?
                        "Loading..." :
                        <div className="description-page container">
                            <div className="description-page__header b-header">
                                <HeaderLogo description={true} handleGoSearchPage={this.handleGoSearchPage} />
                                <HeaderInfo {...this.props.movie}/>
                            </div>
                            <div className="description-page__genre-result">
                                Film by {this.props.genres[0]} genre
                            </div>
                            <MovieItemsBlock
                                movies={this.props.movies}
                            />
                        </div>
                    }
                </React.Fragment>
            </ErrorBoundary>
        )
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(DescriptionPage);
