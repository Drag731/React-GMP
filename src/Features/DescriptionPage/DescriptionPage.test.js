import React from 'react';
import renderer from 'react-test-renderer';
import {shallow, mount, configure} from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import { BrowserRouter } from 'react-router-dom';

import { DescriptionPageTest } from './DescriptionPage'

configure({ adapter: new Adapter() });

describe('<DescriptionPageTest  />', () => {
        const mockProps = {
            genres: [],
            isLoadingMovie: true,
            movie: {},
            movies: [],
            receiveMovieDB: jest.fn(),
            receiveMoviesDB: jest.fn(),
            goToSearch: jest.fn()
        };
        const match = {
            params: {
                id: 1
            }
        };

        it('Renders DescriptionPageTest component', () => {
            const tree = shallow(
                < DescriptionPageTest {...mockProps} match={match} />
            );
            expect(toJson(tree)).toMatchSnapshot();
        });
        it('Renders DescriptionPageTest component snapshot2', () => {
            const tree = shallow(
                < DescriptionPageTest {...mockProps} match={match} isLoadingMovie={false} />
            );
            expect(toJson(tree)).toMatchSnapshot();
        });
        it('should call receiveMovieDB on init', () => {
            const receiveMovieDB = jest.fn();
            const receiveMoviesDB = jest.fn();
            const tree = shallow(
                <DescriptionPageTest
                    {...mockProps}
                    match={match}
                    receiveMovieDB={receiveMovieDB}
                    receiveMoviesDB={receiveMoviesDB}
                />
            );
            expect(receiveMovieDB).toHaveBeenCalled();
            expect(receiveMoviesDB).toHaveBeenCalled();
        });
        it('DescriptionPage gets new props', () => {
            const match1 = {
                params: {
                    id: 1
                }
            };
            const match2 = {
                params: {
                    id: 2
                }
            };
            const changeIsLoading = jest.fn();
            const receiveMovieDB = jest.fn();
            const receiveMoviesDB = jest.fn();
            const tree = mount(
                <DescriptionPageTest
                    {...mockProps}
                    match={match1}
                    receiveMovieDB={receiveMovieDB}
                    receiveMoviesDB={receiveMoviesDB}
                    changeIsLoading={changeIsLoading}
                />
            );

            expect(tree.prop('match')).toEqual(match1);
            tree.setProps({match: match2});
            expect(tree.prop('match')).toEqual(match2);
            expect(changeIsLoading).toHaveBeenCalled();
            expect(receiveMovieDB).toHaveBeenCalled();
            expect(receiveMoviesDB).toHaveBeenCalled();
        });
    it('should call handleGoSearchPage', () => {
        const movie = {
            genres: ['A']
        };
        const tree = mount(
            <BrowserRouter><DescriptionPageTest {...mockProps} match={match} isLoadingMovie={false} movie={movie}/></BrowserRouter>
        );
        const { goToSearch } = mockProps;
        tree.find('.b-header__logo-link').first().simulate('click');
        expect(goToSearch).toHaveBeenCalled();
    });
});
