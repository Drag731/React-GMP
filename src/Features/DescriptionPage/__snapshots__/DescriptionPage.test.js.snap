// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DescriptionPageTest  /> Renders DescriptionPageTest component 1`] = `
<ErrorBoundary>
  <React.Fragment>
    Loading...
  </React.Fragment>
</ErrorBoundary>
`;

exports[`<DescriptionPageTest  /> Renders DescriptionPageTest component snapshot2 1`] = `
<ErrorBoundary>
  <React.Fragment>
    <PageNotFound />
  </React.Fragment>
</ErrorBoundary>
`;
